File: tooling/src/sesame_idam_tooling/verified_principal.py
"""Verified-principal mapping + JSON Schema checks (Epic 15.2 / 15.8)."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from sesame_idam_tooling.oidc_conformance import repo_root_from

PROFILE_VERSION = "1.0.0"
SX_NS = "https://sesame-idam.dev/claims"


def load_schema(root: Path | None = None) -> dict[str, Any]:
    root = root or repo_root_from()
    path = root / "docs" / "standards-first-oidc" / "verified-principal-v1.schema.json"
    return json.loads(path.read_text(encoding="utf-8"))


def map_access_claims_to_principal(claims: dict[str, Any]) -> dict[str, Any]:
    """Map access-token claim dict → verified principal (post-validation)."""
    sx = claims.get(SX_NS) or {}
    roles = sorted(set(sx.get("roles") or []))
    permissions = sorted(set(sx.get("permissions") or []))
    org = claims.get("org_id")
    principal: dict[str, Any] = {
        "profile_version": PROFILE_VERSION,
        "tenant_id": claims["tenant_id"],
        "subject": claims["sub"],
        "client_id": claims["client_id"],
        "application_id": claims["client_id"],
        "session_id": claims["sid"],
        "token_version": int(claims["ver"]),
        "organization_id": org if org is not None else None,
        "user_type": claims["user_type"],
        "roles": roles,
        "permissions": permissions,
        "entitlements_ref": sx.get("entitlements_ref"),
        "entitlements_hash": sx.get("entitlements_hash"),
        "actor": claims.get("act"),
    }
    portal = sx.get("portal")
    if portal:
        principal["portal"] = portal
    return principal


def _type_ok(value: Any, declared: Any) -> bool:
    if isinstance(declared, list):
        return any(_type_ok(value, item) for item in declared)
    if declared == "string":
        return isinstance(value, str)
    if declared == "integer":
        return isinstance(value, int) and not isinstance(value, bool)
    if declared == "array":
        return isinstance(value, list)
    if declared == "object":
        return isinstance(value, dict)
    if declared == "null":
        return value is None
    return True


def validate_principal(principal: dict[str, Any], schema: dict[str, Any] | None = None) -> list[str]:
    """Lightweight draft-2020-12 subset validator for the verified-principal schema."""
    schema = schema or load_schema()
    problems: list[str] = []
    required = schema.get("required") or []
    properties = schema.get("properties") or {}

    for key in required:
        if key not in principal:
            problems.append(f"missing required field {key}")

    if schema.get("additionalProperties") is False:
        for key in principal:
            if key not in properties:
                problems.append(f"unexpected field {key}")

    for key, prop in properties.items():
        if key not in principal:
            continue
        value = principal[key]
        if "const" in prop and value != prop["const"]:
            problems.append(f"{key} must be {prop['const']!r}")
        if "type" in prop and not _type_ok(value, prop["type"]):
            problems.append(f"{key} has wrong type")
        if prop.get("type") == "array" and isinstance(value, list):
            if prop.get("uniqueItems") and len(value) != len(set(value)):
                problems.append(f"{key} must have unique items")
            for item in value:
                if not isinstance(item, str):
                    problems.append(f"{key} items must be strings")
                    break
        if key == "token_version" and isinstance(value, int) and value < 1:
            problems.append("token_version must be >= 1")
    return problems


def sample_access_claims(*, with_org: bool = False) -> dict[str, Any]:
    claims: dict[str, Any] = {
        "sub": "11111111-1111-1111-1111-111111111111",
        "user_id": "11111111-1111-1111-1111-111111111111",
        "client_id": "acme-web",
        "sid": "sid-1",
        "ver": 1,
        "tenant_id": "acme",
        "user_type": "customer",
        SX_NS: {
            "tenant": "acme",
            "portal": "acme-web",
            "roles": ["owner"],
            "permissions": ["org:admin"],
        },
    }
    if with_org:
        claims["org_id"] = "22222222-2222-2222-2222-222222222222"
    return claims

File: microservices/idam/common/src/jwt/verified_principal.rs
//! Access-token claims → verified-principal JSON (Epic 15.2).
//!
//! Mapping rules are normative in
//! `docs/standards-first-oidc/verified-principal-mapping-v1.md`.

use serde_json::{json, Value};

use super::types::AccessClaims;

/// Profile version embedded in [`verified-principal-v1.schema.json`].
pub const VERIFIED_PRINCIPAL_PROFILE_VERSION: &str = "1.0.0";

/// Map validated [`AccessClaims`] into verified-principal v1 JSON.
///
/// Callers must only pass claims produced after cryptographic verification
/// (for example [`super::verify_access_token`]).
pub fn map_access_claims_to_verified_principal(claims: &AccessClaims) -> Value {
    let mut roles = claims.sx.roles.clone();
    roles.sort();
    roles.dedup();
    let mut permissions = claims.sx.permissions.clone();
    permissions.sort();
    permissions.dedup();

    let mut principal = json!({
        "profile_version": VERIFIED_PRINCIPAL_PROFILE_VERSION,
        "tenant_id": claims.tenant_id,
        "subject": claims.sub,
        "client_id": claims.client_id,
        "application_id": claims.client_id,
        "session_id": claims.sid,
        "token_version": claims.ver,
        "organization_id": claims.org_id,
        "user_type": claims.user_type,
        "roles": roles,
        "permissions": permissions,
        "entitlements_ref": claims.sx.entitlements_ref,
        "entitlements_hash": claims.sx.entitlements_hash,
        "actor": claims.act.as_ref().map(|act| json!({
            "sub": act.sub,
        })),
    });

    if !claims.sx.portal.is_empty() {
        principal["portal"] = Value::String(claims.sx.portal.clone());
    }

    principal
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::jwt::builders::{AccessClaimsBuilder, SesameAuthzClaimsBuilder};
    use crate::jwt::signer::Ed25519Signer;
    use crate::jwt::verify_access_token;

    fn sample_claims(with_org: bool) -> AccessClaims {
        let now = chrono::Utc::now().timestamp();
        let sx = SesameAuthzClaimsBuilder::new()
            .tenant("acme")
            .portal("acme-web")
            .roles(vec!["owner".into(), "owner".into()])
            .permissions(vec!["org:admin".into()])
            .build()
            .expect("sx");
        let mut builder = AccessClaimsBuilder::new()
            .iss("https://idam.example.com")
            .sub("11111111-1111-1111-1111-111111111111")
            .aud(vec!["identity-login".into()])
            .client_id("acme-web")
            .scope("openid profile")
            .exp(now + 300)
            .nbf(now - 5)
            .iat(now)
            .jti("jti-vp-1")
            .ver(1)
            .sid("sid-vp-1")
            .tenant_id("acme")
            .user_id("11111111-1111-1111-1111-111111111111")
            .user_type("customer")
            .sx(sx);
        if with_org {
            builder = builder.org_id("22222222-2222-2222-2222-222222222222");
        }
        builder.build().expect("claims")
    }

    #[test]
    fn maps_pre_org_principal_with_null_organization() {
        let principal = map_access_claims_to_verified_principal(&sample_claims(false));
        assert_eq!(principal["profile_version"], "1.0.0");
        assert_eq!(principal["subject"], "11111111-1111-1111-1111-111111111111");
        assert_eq!(principal["session_id"], "sid-vp-1");
        assert_eq!(principal["application_id"], "acme-web");
        assert_eq!(principal["portal"], "acme-web");
        assert_eq!(principal["organization_id"], Value::Null);
        assert_eq!(principal["roles"], json!(["owner"]));
        assert_eq!(principal["permissions"], json!(["org:admin"]));
    }

    #[test]
    fn maps_org_and_validates_against_minted_token() {
        let signer = Ed25519Signer::from_env_or_generate().expect("signer");
        let claims = sample_claims(true);
        let token = signer
            .sign_access_claims(&claims)
            .expect("sign access token");
        let verified = verify_access_token(&signer, &token, Some("acme")).expect("verify");
        let principal = map_access_claims_to_verified_principal(&verified);

        for key in [
            "profile_version",
            "tenant_id",
            "subject",
            "client_id",
            "application_id",
            "session_id",
            "token_version",
            "user_type",
            "roles",
            "permissions",
        ] {
            assert!(principal.get(key).is_some(), "missing required field {key}");
        }
        assert_eq!(
            principal["organization_id"],
            "22222222-2222-2222-2222-222222222222"
        );
        assert_eq!(principal["token_version"], 1);
    }
}
